File: 贪心算法/python/KMP.py
"""
KMP算法用于字符串匹配, 首先约定pat表示模式串, 长度为M. txt表示文本串, 长度为N. KMP算法就是在txt中查找子串pat,
如果存在, 就返回这个子串在txt中的起始索引, 不存在返回-1
"""


def violence_kmp(pat, txt):
    """
    使用暴力算法求解. 时间效率为O(MN), 空间复杂度为O(1)
    使用该解法求pat = "aaab", txt = "aaacaaab", txt中字母"c"被反复比较, 但其实pat中根本不含"c", 所以没必要比较.
    """
    M = len(pat)
    N = len(txt)
    for i in range(N-M+1):
        if txt[i:i+M] == pat:
            return i
    return -1


def kmp(pat, txt):
    """
    kmp算法永不回退txt的指针(不会重复扫描txt), 而是借助dp数组中存储的信息把pat移动到正确的位置继续匹配, 以空间换时间,
    所以时间复杂度为O(N), 空间复杂度为O(M)
    """
    M = len(pat)
    N = len(txt)
    dp_table = {}
    txt_e = list(set(txt))
    for e in set(txt_e):
        dp_table[(0, e)] = 0
    dp_table[(0, pat[0])] = 1
    x = 0
    for j in range(1, M):
        for c in range(len(txt_e)):
            e = txt_e[c]
            dp_table[(j, e)] = dp_table[(x, e)]
        dp_table[(j, pat[j])] = j + 1
        x = dp_table[(x, pat[j])]   # 很关键, 还不是很懂

    j = 0
    for i in range(N):
        j = dp_table[(j, txt[i])]
        if j == M:
            return i - M + 1
    return -1


if __name__ == "__main__":
    pat = "ABABC"
    txt = "CABAABABAC"
    print(kmp(pat, txt))

File: 贪心算法/cpp/KMP.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>

using namespace std;

// 暴力方法创建, O(M^2)
vector<int> construct_table_brute(const string& pat)
{
    int M = pat.size();

    // 构造最长公共前后缀, 这是KMP算法最核心的部分
    vector<int> dp_table(M, -1);
    for(int i=1; i<M; i++)
    {
        string tmp = pat.substr(0, i);
        int j = 0;
        int res = 0;
        while(j<tmp.size()-1)
        {   
            // 从最短的公共前后缀开始校验, 可以从最长的公共前后缀开始校验, 找到了就直接break
            if(tmp.substr(0, j+1) == tmp.substr(tmp.size()-j-1, j+1))
                res = j + 1;
            j++;
        }
        dp_table[i] = res;
    }
    // for(auto& x: dp_table)
    //     cout << x << endl;
    return dp_table;
}

// 迭代创建, O(M)
vector<int> construct_table_iteration(const string& pat)
{
    int M = pat.size();

    vector<int> dp_table(M, -1);

    // 该版本理解起来比较困难, 所以替换
    // int t = -1;
    // int j = 0;
    // while(j < M - 1)
    // {
    //     if(0 > t || pat[j] == pat[t])
    //     {
    //         j++;
    //         t++;
    //         dp_table[j] = t;
    //     }
    //     else
    //         t = dp_table[t];
    // }
    for(int j=1; j < M; j++)
    {   
        int t = dp_table[j-1];
        while(t >= 0 && pat[j-1] != pat[t])
            t = dp_table[t];
        dp_table[j] = t + 1;
    }

    // for(auto& x: dp_table)
    //     cout << x << endl;
    return dp_table;
}

// 效率为O(N)
int KMP(const string& pat, const string& txt)
{
    int M = pat.size();
    int N = txt.size();

    // 构造最长公共前后缀, 这是KMP算法最核心的部分
    vector<int> dp_table = construct_table_iteration(pat);
    
    int j = 0;
    int i = 0;
    while(j < M && i < N)
    {
        if(txt[i] == pat[j] || 0 > j)
        {
            i++;
            j++;
        }
        else
            j = dp_table[j];
    }
    return i - j;   
}

int main()
{
    string pat = "ABABC";
    string txt = "CABAABABCC";
    int res = KMP(pat, txt);
    cout << res << endl;
    cout << txt.substr(res, pat.size()) << endl;
    return 0;
}